### PR TITLE
feat: slideshow

### LINKED ISSUE #16: Introductory slideshow
Basic slideshow as on the Figma design with the following additional items

- [ ] Rotate image every X seconds
- [ ] Priority fetch first image and lazy load the rest
- [ ] Capture left and right arrow key to swipe when hovering mouse over the slideshow

### PR DESCRIPTION
**Pull request by som-04**
Fixes #16

- Image changes every 7 seconds.
- Arrows can be used to change images.
- Sliding animation included.

### CODE REVIEW

File: components/slideshow.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
+'use client';
+import Image from 'next/image';
+import { useCallback, useEffect, useState } from 'react';
+import { FaChevronLeft, FaChevronRight } from 'react-icons/fa';
+
+export default function Carousel({ images }: { images: string[] }) {
```
Comment (GetPsyched) [inline components/slideshow.tsx:6]: Rename to slideshow for consistency.
```suggestion
export default function Slideshow({ images }: { images: string[] }) {
```

File: components/slideshow.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      } else if (event.key === 'ArrowRight') {
+        nextSlide();
+      }
+    },
+    [prevSlide, nextSlide]
+  );
+
+  useEffect(() => {
+    window.addEventListener('keydown', handleKeyDown);
+    return () => window.removeEventListener('keydown', handleKeyDown);
```
Comment (GetPsyched) [inline components/slideshow.tsx:43]: Maybe use `keypress` instead 🤔

File: components/slideshow.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  }, [nextSlide]);
+
+  const handleKeyDown = useCallback(
+    (event: KeyboardEvent) => {
+      if (event.key === 'ArrowLeft') {
+        prevSlide();
+      } else if (event.key === 'ArrowRight') {
+        nextSlide();
+      }
+    },
```
Comment (GetPsyched) [inline components/slideshow.tsx:37]: This should only execute if the slideshow is hovered on.
```suggestion
    (event: KeyboardEvent) => {
      if (!hovered) return;

      if (event.key === 'ArrowLeft') {
        prevSlide();
      } else if (event.key === 'ArrowRight') {
        nextSlide();
      }
    },
```

File: components/slideshow.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+              width={1920}
+              src={image}
+            />
+          </div>
+        ))}
+      </div>
+
+      <button
+        className={`next absolute top-1/2 -translate-y-1/2 transform ${
+          hovered ? 'opacity-100' : 'opacity-0'
```
Comment (GetPsyched) [inline components/slideshow.tsx:88]: Read up on the [group Tailwind class](https://tailwindcss.com/docs/hover-focus-and-other-states#styling-based-on-parent-state) which can be used instead of maintaining a hover state just for adding these classes.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

**Comment by som-04** [comment]
made the above mentioned changes to the code

File: app/[locale]/page.tsx [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import Slideshow from '@/components/slideshow';
+
```
Comment (GetPsyched) [inline app/[locale]/page.tsx:2]: There will be no line break here.

File: components/slideshow.tsx [hunk 6]
```diff
@@ -0,0 +1,94 @@
+'use client';
+import Image from 'next/image';
```
Comment (GetPsyched) [inline components/slideshow.tsx:2]: Add a line break here

File: components/slideshow.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    const intervalId = setInterval(() => {
+      nextSlide();
+    }, 3000);
+
+    return () => clearInterval(intervalId);
+  }, [nextSlide]);
+
+  useEffect(() => {
+    const handleKeyDown = (event: KeyboardEvent) => {
+      if (hovered) {
```
Comment (GetPsyched) [inline components/slideshow.tsx:32]: No, do `if (!hovered) return;`. Doing it this way unnecessarily increases indentation.

File: components/slideshow.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  }, [hovered, prevSlide, nextSlide]);
+
+  return (
+    <div
+      className="relative overflow-hidden"
+      onMouseEnter={() => setHovered(true)}
+      onMouseLeave={() => setHovered(false)}
+    >
+      <button
+        className={`prev absolute top-1/2 z-10 -translate-y-1/2  transform opacity-0 transition-opacity duration-300 hover:opacity-100`}
```
Comment (GetPsyched) [inline components/slideshow.tsx:54]: This is wrong, you're not using the `group-*` syntax which dodges the whole point. Either way, why are you trying to hide the buttons when not hovering on the slideshow anyway? It's not intuitive. Make it `opacity-60` for not hovered and `opacity-100` for hovered. Leave the group thing.

Also, what does the `prev` class do here?

File: components/slideshow.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+              loading={index === 0 ? 'eager' : 'lazy'}
+              width={1920}
+              src={image}
+            />
+          </div>
+        ))}
+      </div>
+
+      <button
+        className={`next absolute top-1/2 z-10 -translate-y-1/2 transform  opacity-0 transition-opacity duration-300 hover:opacity-100`}
```
Comment (GetPsyched) [inline components/slideshow.tsx:86]: Same here

File: components/slideshow.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        onClick={nextSlide}
+        style={{ right: '16px' }}
```
Comment (GetPsyched) [inline components/slideshow.tsx:88]: Huh? Use Tailwind for all styling.

File: components/slideshow.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+      >
+        <FaChevronRight className="text-white" style={{ fontSize: '40px' }} />
```
Comment (GetPsyched) [inline components/slideshow.tsx:90]: All react icons have a `size` prop, use that.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: components/slideshow.tsx (hunk 8 above)
Comment (som-04) [reply components/slideshow.tsx:54]: Oh I see, so that way the buttons will be in focus the moment we hover the slideshow and not the buttons themselves.

File: app/[locale]/page.tsx (hunk 5 above)
Comment (som-04) [reply app/[locale]/page.tsx:2]: unable to commit without the line break
image

**Comment by som-04** [comment]
- had to use in-line CSS for dynamic calculation of styles apart from that all is tailwind.
- used `group` syntax.
- used `if(!hover) return;`
- made the line break changes except for the `page.tsx` file.

File: app/[locale]/page.tsx (hunk 5 above)
Comment (GetPsyched) [reply app/[locale]/page.tsx:2]: Hmm, then change the below imports to use `@/app/directors-corner` syntax and then you can remove the line break

File: components/slideshow.tsx [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    };
+    window.addEventListener('keydown', handleKeyDown);
+    return () => {
+      window.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [hovered, prevSlide, nextSlide]);
+
+  return (
+    <div
+      className="group relative overflow-hidden"
```
Comment (GetPsyched) [inline components/slideshow.tsx:48]: Don't panic, read my previous comment carefully,

> Make it `opacity-60` for not hovered and `opacity-100` for hovered. Leave the group thing.

Because with the group tag, the hover will apply when you hover on the div itself; which is wrong. You change the opacity only when hovering the arrow icon.

I originally suggested to use group because I didn't notice that you did it wrong initially.

**Review by GetPsyched — CHANGES_REQUESTED** [review]
Refer [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element) to make the HTML tags semantic. Won't merge `div` tags.

**Comment by GetPsyched** [comment]
Please mark previous things as resolved as you resolve them.

File: components/slideshow.tsx (hunk 2 above)
Comment (som-04) [reply components/slideshow.tsx]: wasn't working with keypress

File: components/slideshow.tsx (hunk 12 above)
Comment (som-04) [reply components/slideshow.tsx:48]: oh my bad

File: components/slideshow.tsx (hunk 2 above)
Comment (GetPsyched) [reply components/slideshow.tsx]: Okay, will test this locally.

File: components/slideshow.tsx [hunk 13]
```diff
@@ -66,15 +77,20 @@ export default function Slideshow({
             )}
           </figure>
         ))}
-      </section>
-
-      <section className="container absolute inset-0 flex min-w-full justify-between">
-        <button className="opacity-50 hover:opacity-100" onClick={prevSlide}>
-          <FaChevronLeft className="my-auto text-neutral-100" size={40} />
+      </Slider>
+      <section className="absolute inset-0 flex items-center justify-between px-20">
```
Comment (GetPsyched) [inline components/slideshow.tsx:81]: Why did you remove the container class here?

File: components/slideshow.tsx [hunk 14]
```diff
@@ -1,53 +1,64 @@
 'use client';
-
+import React, { useEffect, useRef } from 'react';
```
Comment (GetPsyched) [inline components/slideshow.tsx:2]: What's this diff? Why was the extra line removed?

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: components/slideshow.tsx
Comment (GetPsyched) [inline components/slideshow.tsx:1]: If it wasn't responding to the useEffect, then why not remove it?

**Review by GetPsyched — CHANGES_REQUESTED** [review]

File: components/slideshow.tsx [hunk 15]
```diff
@@ -0,0 +1,80 @@
+'use client';
+
+import React, { useRef } from 'react';
+import Image from 'next/image';
+import { FaChevronLeft, FaChevronRight } from 'react-icons/fa';
+import Slider from 'react-slick';
+import 'slick-carousel/slick/slick.css';
+import 'slick-carousel/slick/slick-theme.css';
```
Comment (GetPsyched) [inline components/slideshow.tsx:8]: If using VSCode, right click > source action > organise imports.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

**Review by GetPsyched — CHANGES_REQUESTED** [review]
We might need to revert adding `react-slick` as it makes things very laggy.
Peek 2024-03-03 21-37

**Comment by som-04** [comment]
I'll try it with react responsive carousel.

**Comment by GetPsyched** [comment]
Try it with ShadCN too; spoke with @geekysilento, it might be worth it.

**Comment by som-04** [comment]
used shadcn except the arrow keys working upon hover everything is pretty much same as before.
the only problem is that the arrow keys work once the user clicks on the buttons present there.

**Comment by som-04** [comment]
wasn't letting the commit through without adding these line breaks in `slideshow.tsx` and `carousel.tsx` respectively

image
image

File: package.json [hunk 16]
```diff
@@ -19,21 +19,28 @@
...
+    "embla-carousel-autoplay": "^8.0.0",
+    "embla-carousel-react": "^8.0.0",
     "negotiator": "^0.6.3",
     "next": "14.0.4",
     "next-auth": "^4.24.5",
     "react": "^18",
     "react-dom": "^18",
     "react-icons": "^5.0.1",
+    "react-slick": "^0.30.2",
+    "slick-carousel": "^1.8.1",
```
Comment (GetPsyched) [inline package.json:35]: You need to remove these.

**Review by GetPsyched — CHANGES_REQUESTED** [review]

**Comment by GetPsyched** [comment]
> wasn't letting the commit through without adding these line breaks in `slideshow.tsx` and `carousel.tsx` respectively

Yes, that is intended. The line breaks are enforced by the following rule:
https://github.com/nitkkr-dev/nitkkr/blob/6a9011435761f48bc972a123bf4d7471b56499dd/.eslintrc.json#L45

**Comment by GetPsyched** [comment]
Huh, you moved the attributes from `<figure>`, not `<article>`. `<figure>` is needed there, how else are you using `<figcaption>`

**Comment by som-04** [comment]
oh sorry mb lol overlooked that

**Comment by som-04** [comment]
ono there are run time errors will fix them

**Review by GetPsyched — APPROVED** [review]
LGTM
